### ISSUE TITLE
Java: got "1.943037616030839E16" for 1.9430376160308388E16 input (even rounding mode)

### ISSUE THREAD

**Issue by plokhotnyuk** [ulfjack/ryu#29]
```
IN=100001101010001010000011111010010111111001110001100101100101001
   S=+ E=2 M=4857594040077097
E =16
d+=19430376160308390
d =19430376160308388
d-=19430376160308386
e2=0
19430376160308388 * 2^0
V+=19430376160308390
V =19430376160308388
V-=19430376160308386
19430376160308388 2
d+=19430376160308390
d =19430376160308388
d-=19430376160308386
e10=0
d-10=false
LAST_REMOVED_DIGIT=8
VP=1943037616030839
VR=1943037616030838
VM=1943037616030838
O=1943037616030839
OLEN=16
EXP=16
1.943037616030839E16 1.9430376160308388E16
```

**Comment by plokhotnyuk** [comment]
Here is list of 28 wrong results for 10^7 pseudo-random inputs:
```
in: -6.9741824662760956E19 out: -6.974182466276096E19
in: -5.8906602102955516E19 out: -5.890660210295552E19
in: 4.4236755337039356E19 out: 4.423675533703936E19
in: 4.2695312359520957E18 out: 4.269531235952096E18
in: -1.3385859731115519E20 out: -1.338585973111552E20
in: 6.4490693738698236E19 out: 6.449069373869824E19
in: -5.5233239189231995E18 out: -5.5233239189232E18
in: -2.4467899820140477E18 out: -2.446789982014048E18
in: -2.5835890667033277E18 out: -2.583589066703328E18
in: 3.4006137342603197E18 out: 3.40061373426032E18
in: -5.0229339435657595E18 out: -5.02293394356576E18
in: 3.7816698400594877E18 out: 3.781669840059488E18
in: 3.2114673323227837E18 out: 3.211467332322784E18
in: -2.3790871870256317E18 out: -2.379087187025632E18
in: -2.8997671660051198E19 out: -2.89976716600512E19
in: 4.3816050601147837E18 out: 4.381605060114784E18
in: 4.0185858599547837E18 out: 4.018585859954784E18
in: 1.5898203054348159E19 out: 1.589820305434816E19
in: 3.5020980314022077E18 out: 3.502098031402208E18
in: 2.8077660000142078E19 out: 2.807766000014208E19
in: 3.3420968774834877E18 out: 3.342096877483488E18
in: 9.716820539547519E18 out: 9.71682053954752E18
in: 1.3064592373011839E19 out: 1.306459237301184E19
in: 2.8651069053029118E19 out: 2.865106905302912E19
in: 4.4552840615267517E18 out: 4.455284061526752E18
in: -1.8438980310768638E20 out: -1.843898031076864E20
in: 2.6056580893472957E18 out: 2.605658089347296E18
in: -2.5895306871597757E18 out: -2.589530687159776E18
```

**Comment by plokhotnyuk** [comment]
It looks that this regression was introduced in latest couple commits. The the initial version serialize it without rounding.

**Comment by ulfjack** [comment]
Thanks for the report. It was due to a typo I introduced when I translated some recent-ish performance improvements from C to Java. Oddly enough, it wasn't discovered by my extensive comparison test - maybe that generates too predictable patterns.

**Comment by ulfjack** [comment]
I must have been asleep when I did that translation - I found another bug and fixed it in 6ae35dcad2d85fed7c118d37210d914fc77377e7. I also added a randomized comparison test, so hopefully, this will not occur again.
